core/src/main/scala/org/apache/spark/deploy/history/ApplicationCache.scala

```diff
@@ -58,24 +58,26 @@ private[history] class ApplicationCache(
 
   }
 
-  private val removalListener = new RemovalListener[CacheKey, CacheEntry] {
+  private val cacheWriter = new CacheWriter[CacheKey, CacheEntry] {
```

Comment:
`CacheWriter ` adopts sync remove behavior similar to guava and `RemovalListener ` always Asynchronous

Reply:
fyi, `CacheWriter` will be deprecated and replaced in 2.9, and removed in 3.0. Instead `Caffeine.evictionListener(RemovalListener)` will provide the sync remove behavior, and any other atomic writes can be captured manually via `asMap().compute` methods. Should be minor change for you when 2.9 is released.

Reply:
@ben-manes Thank you very much for your advice.

Reply:
@ben-manes I try to use `Caffeine.evictionListener(RemovalListener) ` instead of `CacheWriter`,  but it seems that the `invalidate` method does not trigger the execution of `RemovalListener#onRemoval`?

Reply:
@LuciferYang This can be accomplished by using `cache.asMap().computeIfPresent` instead of `invalidate`. Then you may callback manually within the atomic operation and return `null` to remove the entry.

`CacheWriter` tried to intercept but it was confusing by missing some cases. For example should a `invalidate(key)` call the writer if the entry is absent? It does not, but in a multi-layer cache one might expect to propagate the removal even if absent. These quirks led to replacing it with explicit callbacks via compute for manual operations and `evictionListener` for automatic ones. An invalidation is a manual operation, so it requires a map compute.